nixos/redshift: use config file instead of passing parameters, using RFC42

thiagokokada opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Not every option is exposed by redshift/gammastep parameters, for example gamma options are only exposed in configuration file. So this PR refactors this module to generate a configuration file and pass it to the redshift/gammastep using -c parameter.

This is a breaking change since there is no support for some of the older options like `extraOptions`. Also, the way to pass settings completely changed (but there is some aliases).

Second try of PR #108625, now following [RFC 42](https://github.com/NixOS/rfcs/pull/42). Also some parts of it taken from PR #50979.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

thiagokokada committed:
nixos/redshift: use config file instead of passing parameters

Not every option is exposed by redshift/gammastep parameters, for
example gamma options are only exposed in configuration file. So this PR
refactors this module to generate a configuration file and pass it to
the redshift/gammastep using -c parameter.

This is a breaking change since there is no support for some of the
older options like `extraConfig`. Also, the way to pass settings
completely changed (but there is some aliases).

thiagokokada commented:
Some tests:

From this config
```nix
{
    location.provider = "geoclue2";
    services.redshift = {
      enable = true;
      settings = {
        general = {
          temp-day = 5500;
          temp-night = 3700;
          gamma = 0.8;
          fade = 1;
        };
      };
      package = pkgs.unstable.gammastep;
      executable = "/bin/gammastep-indicator";
    };
}
```
Result `redshift.conf` file:
```
[general]
gamma=0.800000
temp-day=5500
temp-night=3700
```

Backwards compatibility:

```nix
{
    location = {
      provider = "manual";
      latitude = -23.3;
      longitude = -46.6;
    };

    redshift = {
      enable = true;
      temperature = {
        day = 5500;
        night = 3700;
      };
      brightness = {
        day = 1;
        night = 0.8;
      };
    };
}
```

Result `redshift.conf` file:
```
[manual]
lat=-23.300000
lon=-46.600000

[redshift]
brightness-day=1
brightness-night=0.800000
temp-day=5500
temp-night=3700
```

thiagokokada commented:
CC @aanderse @xaverdh @sumnerevans @Infinisil @hyperfekt

thiagokokada commented on nixos/modules/services/x11/redshift.nix:
```diff
@@ -91,48 +63,77 @@ in {
-      example = [ "-v" "-m randr" ];
+    settings = mkOption {
+      type = types.submodule {
+        freeformType = with types; (attrsOf (attrsOf (oneOf [str int float])));
```
Not adding `bool` here since the only bool option in documentation is `fade`, and it would need to be converted from `true`/`false` to `1`/`0`.

thiagokokada replied:
Maybe `settingsFormat.types` already does this conversion :thinking: ?

thiagokokada replied:
But `settingsFormat.types` seems to support `null` too, that I don't think it is desirable here.

Infinisil replied:
One of the main reasons these formats have been abstracted into `pkgs.formats` is that you won't have to encode the type of them yourself. INI supports booleans, integers and nulls, but there's no need to worry about that since all the valid types and their string generation is handled by the `generator` function, which matches what its `type` can represent. So I suggest just using `settingsFormat.generate` here. This way the module can be reused for future redshift versions, even if that newly uses booleans and nulls in its config.

Note however that redshift uses integers 0/1 to encode `fade`, not booleans. So even though the Nix `type` does allow booleans you'll still have to use 0/1 to encode `fade`.

thiagokokada replied:
@Infinisil So I remove the `type` here or change it to `type = types.submodule { freeFormType = settingsFormat.types; }` like I saw in RFC42 :thinking: ?

Asking because I am already using `settingsFormat.generate` (I may be using it wrong though): https://github.com/NixOS/nixpkgs/pull/108739/files#diff-61e7d28082c9dd1fe9bd0282f55b1cd33f1397fa87b424b6fdf71253e591bc02R122

thiagokokada replied:
BTW, the reason I decided to overwrite the types is to avoid some typos in the config, like the mentioned `fade` case already or someone trying to pass something as `null` to remove a default config (AFAIK this will not happen). And `redshift` doesn't have a release for almost 3 years already, so I think they breaking their config now is really unlikely.

But I think it is fair enough to use the more generic types too.

thiagokokada replied:
If the tests implemented now it was easy to check what works and what doesn't.

thiagokokada commented on nixos/modules/services/x11/redshift.nix:
```diff
@@ -23,7 +26,14 @@ in {
           else builtins.fromJSON value))
     (mkRenamedOptionModule [ "services" "redshift" "provider" ] [ "location" "provider" ])
-  ];
+    (mkRemovedOptionModule [ "services" "redshift" "extraOptions" ] "All Redshift configuration is now available through services.redshift.settings instead.")
+  ] ++
+  (map (option: mkRenamedOptionModule ([ "services" "redshift" ] ++ option.old) [ "services" "redshift" "settings" "redshift" option.new ]) [
+      { old = [ "temperature" "day" ];    new = "temp-day"; }
+      { old = [ "temperature" "night" ];  new = "temp-night"; }
+      { old = [ "brightness" "day" ];     new = "brightness-day"; }
+      { old = [ "brightness" "night" ];   new = "brightness-night"; }
+    ]);
```
Took from: https://github.com/NixOS/nixpkgs/pull/50979

thiagokokada commented on nixos/modules/services/x11/redshift.nix:
```diff
@@ -91,48 +63,77 @@ in {
-        else lcfg.provider;
-    in
-    {
+    assertions = [ {
+      assertion = (cfg.settings ? ${mainSection}.dawn-time) == (cfg.settings ? ${mainSection}.dusk-time);
+        message = "Time of dawn and time of dusk must be provided together.";
+    } ];
+
+    services.redshift.settings.manual = mkIf isManual {
+      lat = mkIf options.location.latitude.isDefined lcfg.latitude;
+      lon = mkIf options.location.longitude.isDefined lcfg.longitude;
+    };
```
Took from: https://github.com/NixOS/nixpkgs/pull/50979

thiagokokada commented on nixos/modules/services/x11/redshift.nix:
```diff
@@ -91,48 +63,77 @@ in {
+      # https://gitlab.com/chinstrap/gammastep/-/commit/1608ed61154cc652b087e85c4ce6125643e76e2f
+      then "general"
+      else "redshift";
+    target = if (cfg.settings.${mainSection}.adjustment-method or null) == "drm"
+      then "basic.target"
+      else "graphical-session.target";
```
Took from: https://github.com/NixOS/nixpkgs/pull/50979

thiagokokada commented on nixos/modules/services/x11/redshift.nix:
```diff
@@ -91,48 +63,77 @@ in {
         '';
         RestartSec = 3;
-        Restart = "always";
+        Restart = "on-failure";
```
Avoid restart loop in case of configuration failure.

thiagokokada commented:
Whoops, missed adding the provider to config.

thiagokokada commented:
Done, using the first `/etc/nixos/configuration.nix` from here https://github.com/NixOS/nixpkgs/pull/108739#issuecomment-756484898:

```
[general]
fade=1
gamma=0.800000
location-provider=geoclue2
temp-day=5500
temp-night=3700
```

thiagokokada commented on nixos/modules/services/x11/redshift.nix:
```diff
@@ -88,51 +69,90 @@ in {
       example = "/bin/redshift-gtk";
```
Since we are already breaking the backwards compatibility with this module I am thinking of changing this option to:

```nix
{
  type = types.path;
  default = ${pkgs.redshift}/bin/redshift;
}
```

This way, we at least have some type checking in here, but on the other hand this may make this module more error prone since the user may forget to set the correct `cfg.package` option (that is still necessary for https://github.com/NixOS/nixpkgs/pull/108739/files#diff-61e7d28082c9dd1fe9bd0282f55b1cd33f1397fa87b424b6fdf71253e591bc02R125).

If someone has a better solution, suggestions are welcome too.

thiagokokada replied:
Add an assertion that checks if the executable actually exists. This should fix most issues.

thiagokokada committed:
nixos/redshift: add tests

thiagokokada commented:
Commit 39c1ce1 add tests so we can make sure this module works.

thiagokokada committed:
nixos/redshift: add thiagokokada as maintainer

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/437

thiagokokada commented:
@GrahamcOfBorg build opentabletdriver

thiagokokada commented:
@GrahamcOfBorg build redshift

thiagokokada commented:
@GrahamcOfBorg test redshift

thiagokokada commented:
/marvin opt-in
/status needs_reviewer

Infinisil commented on nixos/modules/services/x11/redshift.nix:
```diff
@@ -88,51 +71,101 @@ in {
-  config = mkIf cfg.enable {
+  config = let
+    mainSection = if strings.hasInfix "gammastep" cfg.executable
+      # https://gitlab.com/chinstrap/gammastep/-/commit/1608ed61154cc652b087e85c4ce6125643e76e2f
+      then "general"
+      else "redshift";
```
Not needed anymore with the type change suggested above

Infinisil commented on nixos/modules/services/x11/redshift.nix:
```diff
@@ -88,51 +71,101 @@ in {
-      default = [];
-      example = [ "-v" "-m randr" ];
+    settings = mkOption {
+      type = types.submodule {
+        freeformType = settingsFormat.type;
+      };
```
```suggestion
      type = types.submodule ({ options, ... }: {
        freeformType = settingsFormat.type;

        options.redshift = mkOption {
          type = settingsFormat.type.functor.wrapped;
        };

        # Copy all redshift definitions to general, such that in case of gammastep,
        # which prefers general over redshift, gets both values from general and redshift
        # While the original redshift just gets values from redshift, as it should
        config.general = lib.modules.mkAliasAndWrapDefsWithPriority id options.redshift;
      });
```

This then obviates the need for the `mainSection` hack below. You can then just use `redshift` for the definitions to be picked up by both redshift and gammastep.

thiagokokada replied:
I don't know if copying would work. `gammastep` actually supports both `redshift` and `general` sections (but emits a warning in case `redshift` section is found).

I can tests if this works, but still looks wrong.

thiagokokada replied:
Also, I am quite sure that `redshift` actually checks for invalid inputs and gives an error in case of them being set. But let me try first and I will post the results later on.

Infinisil replied:
Oh, yeah my assumption was that redshift ignores the `general` section, and gammastep ignores the `redshift` section if `general` exists. Knowing the behavior exactly would be great to find a good solution here.

thiagokokada replied:
Well, this actually works TIL. Ok, I think it is fair, going to do like you said here.

Infinisil commented on nixos/modules/services/x11/redshift.nix:
```diff
@@ -88,51 +71,101 @@ in {
+      # https://gitlab.com/chinstrap/gammastep/-/commit/1608ed61154cc652b087e85c4ce6125643e76e2f
+      then "general"
+      else "redshift";
+    target = if (cfg.settings.${mainSection}.adjustment-method or null) == "drm"
```
With above type change, this should check the value for `cfg.settings.general.adjustment-method` (since this will be set whether the user sets `general` or `redshift`.

Infinisil commented on nixos/modules/services/x11/redshift.nix:
```diff
@@ -88,51 +71,101 @@ in {
-    in
-    {
+    assertions = [
+      {
+        assertion = (cfg.settings ? "redshift") != (cfg.settings ? "general");
+        message = ''Both [redshift] and [general] sections found in settings. For gammastep use "general = {}" instead of "redshift = {}" in "server.redshift.settings".'';
+      }
```
Not necessary with above type change

Infinisil commented on nixos/modules/services/x11/redshift.nix:
```diff
@@ -88,51 +71,101 @@ in {
+        assertion = (cfg.settings ? "redshift") != (cfg.settings ? "general");
+        message = ''Both [redshift] and [general] sections found in settings. For gammastep use "general = {}" instead of "redshift = {}" in "server.redshift.settings".'';
+      }
+      {
+        assertion = (pathExists fullExecutable);
+        message = ''Executable not found. Are you using "/bin/redshift" for "pkgs.redshift" or "/bin/gammastep" for "pkgs.gammastep"?'';
+      }
```
This should be removed, because checking whether a derivation path exists can only be done after the derivation is built, meaning in order to evaluate such a NixOS configuration, Nix needs to first evaluate to this point, then build the derivation, then evaluate again, which should be avoided because it's slow.

Infinisil commented on nixos/modules/services/x11/redshift.nix:
```diff
@@ -88,51 +71,101 @@ in {
+        message = ''Executable not found. Are you using "/bin/redshift" for "pkgs.redshift" or "/bin/gammastep" for "pkgs.gammastep"?'';
+      }
+      {
+        assertion = (cfg.settings ? ${mainSection}.dawn-time) == (cfg.settings ? ${mainSection}.dusk-time);
```
Should also use the `general` section here with above type change.

thiagokokada replied:
I will use `redshift` just to keep one standard (i.e.: avoid errors in future in case someone needs to change this module).

Infinisil commented on nixos/modules/services/x11/redshift.nix:
```diff
@@ -88,51 +71,101 @@ in {
+    ];
+
+    services.redshift.settings = {
+      ${mainSection}.location-provider = lcfg.provider;
```
This however should use the `redshift` name, since that will be copied to both sections.

Infinisil commented on nixos/modules/services/x11/redshift.nix:
```diff
@@ -88,51 +71,101 @@ in {
+      ${mainSection}.location-provider = lcfg.provider;
+      manual = mkIf isManual {
+        lat = mkIf options.location.latitude.isDefined lcfg.latitude;
+        lon = mkIf options.location.longitude.isDefined lcfg.longitude;
```
Don't check for them being defined here, because we want an error if they're not defined.

thiagokokada replied:
No we don't. It is perfectly possible to use provider = `manual` and set `dawn-time` and `dusk-time` instead (this is what I do.

thiagokokada replied:
Can location provider be set to `null` :thinking: ? If yes,  I think I can remove those `mkIf`s for latitude/longitude.

thiagokokada replied:
Hum... Maybe checking if not `dawn-time` and `dusk-time` is set too would be better.

Infinisil replied:
Ah so by selecting `manual` you have to either use `lat`/`lon` OR `dawn-time`/`dusk-time`? What happens if both are set?

thiagokokada replied:
`dawn-time`/`dusk-time` has priority, I just think it is silly to make the user must set `latitude`/`longitude` when they want to use `dawn-time`/`dusk-time`.

thiagokokada replied:
What I am thinking is a check like this:

```nix
manual = mkIf isManual && (cfg.settings ? ${mainSection}.dawn-time) && (cfg.settings ? ${mainSection}.dusk-time)
```

It is quite big, but I think would work. WDYT @Infinisil :thinking: ?

thiagokokada replied:
Well, this doesn't work, infinite recursion.

thiagokokada replied:
Ok, going to remove the `mkIf`s. Can't think of a better solution and setting `location = { latitude = 0; longitude = 0; }` does the trick.

Infinisil replied:
I'll take a look at this again later, but this isn't a great solution at the moment. There's a mismatch between redshift's idea of "manual" and the location provider's idea of "manual". I think the location provider module might need some change.

thiagokokada replied:
I also don't have a problem of having this merge as is and after those changes in `location` is done I end up fixing this. Like, it is not ideal to have to set location to use time options but it works.

BTW, I think `location` should support a `location.provider = null`. This would work in this case because if provider isn't set we just use full manual mode.

Infinisil commented on pkgs/applications/misc/redshift/default.nix:
```diff
@@ -7,6 +7,7 @@
 , withRandr ? stdenv.isLinux, libxcb
 , withDrm ? stdenv.isLinux, libdrm
 
+, nixosTests
```
I guess you meant to add this to passthru?

thiagokokada replied:
I will remove it. I tried to add this to `passthru.tests` but since redshift is have 3 different derivations and my tests doesn't support it, I decided for now to not add the tests to `passthru.tests`.

thiagokokada commented:
@Infinisil Fixed the issues, rebased and fix the merge conflicts. Can you review again?

thiagokokada commented:
@GrahamcOfBorg build redshift

thiagokokada commented:
@GrahamcOfBorg test redshift

thiagokokada commented:
Just fixing a typo in commit.

Infinisil commented:
As far as I know, for redshift to work it needs either:
- `{dawn,dusk}-time` set
- `location-provider = geoclue2` and geoclue enabled
- `location-provider = manual` and `manual.{lon,lat}` set

With the current module the user would have a very hard time knowing about this. So I think let's go for these options instead:
- `services.redshift.useGeoclue`, by default `false`. If `true`, enable geoclue, and set `settings.location-provider = "geoclue2"`
- `services.redshift.{dawn,dusk}Time`, both `null` by default. If non-null (with assertion to verify both are set at once) and `useGeoclue` is false (needed only here because `-time` overrides `location-provider`), set the values in `settings.{dawn,dusk}-time`
- `services.redshift.{longitude,latitude}`, both `null` by default. If non-null (with assertion to verify both are set at once), set the values in `settings.manual.{lon,lat}`

Finally, an assertion that checks ensures one of these is true:
- `settings ? {dawn,dusk}-time`
- `settings.location-provider or "" == "geoclue2"
- `settings.location-provider or "" == "manual" && settings ? manual.{lat,lon}`

If that's not the case, it throws an error, pointing the user to the `useGeoclue`, `{dawn,dusk}Time` and `{longitude,latitude}` options.

Unfortunately we can't easily integrate all this into the `settings` option directly, because we will need to distinguish between options being set and unset, and we don't want `null` to end up in the INI file.

Finally, for backwards compat, if `location.provider` is `"manual"`, we propagate `location.{longitude,latitude}` to `services.redshift.{longitude,latitude}`

thiagokokada commented:
@Infinisil Some questions:

> Unfortunately we can't easily integrate all this into the settings option directly, because we will need to distinguish between options being set and unset, and we don't want null to end up in the INI file.

Can't we use `options.redshift.<option>.isDefined` to check if it exists :thinking: ? This way we wouldn't need to add non-settings options.

> Finally, an assertion that checks ensures one of these is true:

I think this assertion seems really fair. I think (maybe I am wrong) that I can implement it without changing the current code too much. I will give a try later on.

Infinisil commented:
> Can't we use `options.redshift.<option>.isDefined` to check if it exists thinking ? This way we wouldn't need to add non-settings options.

Yeah, but the `format.generate` function doesn't do that check before accessing all the values from `cfg.settings`, and it would be annoying to first filter out all unset values. I'll be thinking more about if this can be done better in the future, as an extension of freeform modules or the `pkgs.formats`.

thiagokokada commented:
@Infinisil Commit a1701cd should cover all user cases. Tested:

```nix
{
      services.redshift = {
        enable = true;
        settings = {
          redshift = {
            temp-day = 1500;
            temp-night = 1500;
            dawn-time = "6:00-7:45";
            dusk-time = "18:35-20:15";
            gamma = 0.8;
            fade = 0;
          };
        };
      };
}
```

```nix
{
      location = {
        provider = "manual";
        latitude = 0.0;
        longitude = 0.0;
      };

      services.redshift = {
        enable = true;
        settings = {
          redshift = {
            temp-day = 1500;
            temp-night = 1500;
            gamma = 0.8;
            fade = 0;
          };
        };
      };
}
```

```nix
{
      location = {
        provider = "geoclue2";
      };

      services.redshift = {
        enable = true;
        settings = {
          redshift = {
            temp-day = 1500;
            temp-night = 1500;
            gamma = 0.8;
            fade = 0;
          };
        };
      };
}
```

Also these fails as excepted:

```nix
{
      services.redshift = {
        enable = true;
        settings = {
          redshift = {
            temp-day = 1500;
            temp-night = 1500;
            dawn-time = "6:00-7:45";
            gamma = 0.8;
            fade = 0;
          };
        };
      };
}
```

```nix
      services.redshift = {
        enable = true;
        settings = {
          redshift = {
            temp-day = 1500;
            temp-night = 1500;
            gamma = 0.8;
            fade = 0;
          };
        };
      };
```
```

thiagokokada committed:
nixos/redshift: add dawnTime/duskTime/latitude/longitude/useGeoclue options

This should make it more discoverable for the user what is the available
options to needed to make the module work.

thiagokokada commented:
@GrahamcOfBorg Returned the following error:

```
nix-env did not evaluate cleanly:
 ["trace: warning: Option `services.redshift.settings.redshift\' has no description."]
```

Added a basic description in commit 791a14f.

infinisil reviewed (changes requested):
Quoting myself [from yesterday](https://github.com/NixOS/nixpkgs/pull/108739#issuecomment-759886533):

> With the current module the user would have a very hard time knowing about this.

The intention being that the option listing should give a good idea of how the module can be configured for common use cases. This is also [mentioned in the RFC](https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md#part-2-1).

And this is the reason why I wish to have a design as described in the linked comment, it allows the users to discover the `services.redshift.{useGeoclue,{dusk,dawn}Time,longitude,latitude}` options.

thiagokokada commented:
@Infinisil I think it is pretty clear that the user needs to check `man 1 redshift` or `man 1 gammastep` to discover the options. IMO, the examples make it quite clear on how to proceed from here on, and the asserts are there only to avoid silly mistakes.

> The intention being that the option listing should give a good idea of how the module can be configured for common use cases. This is also mentioned in the RFC.

If we really think about it the most common usage of redshift can be argued with either location set explicitly or using Geoclue2. `dawn-time` and `dusk-time` does seems to me quite niche cases.

But if this is the only way to have this PR merged, ok, I will do it as you described. But I still don't concur with all the reasons described in it.

thiagokokada commented:
@Infinisil Can't we at least drop `services.redshift.{longitude,latitude}` and `services.redshift.useGeoclue`? I think this will make this module really complicated because:

- To keep backwards compatibility with them, I need to copy `lcfg.{longitude,latitude,provider}` only if not set by `redshift`
- Or we can forget backwards compatibility and make a clean break, but this way this module will be in a really weird state since other similar modules use `location` (like `clight`)

I am really having difficult to fix this module the way it is described in https://github.com/NixOS/nixpkgs/pull/108739#issuecomment-759886533, because things get messy relatively quickly.

One of the reasons that this module doesn't have any described options is because `location` settings does mostly of the heavy lifting. Maybe I could redirect the user for that documentation instead.

I think implementing only `dawnTime` and `duskTime` seems easy enough, but we may go back to the point if those are really that popular options that needs a special handling.

thiagokokada commented:
@Infinisil Give it a try in commit 31ff34b. It is kinda messy but I think it should look ok enough.

thiagokokada commented on nixos/modules/services/x11/redshift.nix:
```diff
@@ -1,29 +1,27 @@
-  ];
+    (mkRemovedOptionModule [ "services" "redshift" "extraOptions" ] "All Redshift configuration is now available through services.redshift.settings instead.")
+  ] ++
+  (map (option: mkRenamedOptionModule ([ "services" "redshift" ] ++ option.old) [ "services" "redshift" "settings" "redshift" option.new ]) [
+      { old = [ "temperature" "day" ];    new = "temp-day"; }
+      { old = [ "temperature" "night" ];  new = "temp-night"; }
+      { old = [ "brightness" "day" ];     new = "brightness-day"; }
+      { old = [ "brightness" "night" ];   new = "brightness-night"; }
+    ]);
```
Since we are already adding some options, maybe I could simply get those options back too with their old descriptions.

WDYT @Infinisil ?

thiagokokada commented on nixos/modules/services/x11/redshift.nix:
```diff
@@ -32,53 +30,74 @@ in {
+    latitude = mkOption {
+      type = types.nullOr types.float;
+      default = null;
+      example = "0.0";
```
I don't know if this is a bug, but setting it as a float was making it not appear in manual.

thiagokokada commented:
@Infinisil Sorry to bother, but can you re-review this? I think this is now what you requested in comment https://github.com/NixOS/nixpkgs/pull/108739#issuecomment-759886533.

Infinisil commented:
I don't have a lot of time recently sorry. The redshift module unfortunately is rather complicated and some changes are still needed to make it work nicely. I made some initial ones [here](https://github.com/Infinisil/nixpkgs/commit/b27fb70444c8db3fa257ef8874aec491bea5b349), but I'm out of brain power for today, I hope to have time for some clean module writing soon!

thiagokokada commented:
@Infinisil Good to know that this PR will be in good hands :smile: . I kinda of want of it merged ASAP, but since I use the release channels and there is still sometime until 21.05 it probably can wait.

xaverdh commented:
I tested this with my configuration, looks good so far.

thiagokokada commented:
/marvin opt_out

thiagokokada commented:
If there is someone interested in this PR I can rebase it to a newer release, however I will probably remove myself as maintainer since I don't use it anymore (I am using the equivalent from Home-Manager).

thiagokokada commented:
Closing some old PRs of mine.
